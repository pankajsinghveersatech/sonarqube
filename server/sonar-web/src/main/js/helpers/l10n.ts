/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { fetchL10nBundle } from '../api/l10n';
import { L10nBundle, L10nBundleRequestParams } from '../types/l10n';
import { toNotSoISOString } from './dates';
import { get as loadFromLocalStorage, save as saveInLocalStorage } from './storage';

export type Messages = T.Dict<string>;

export const DEFAULT_LOCALE = 'en';
export const DEFAULT_MESSAGES = {
  // eslint-disable-next-line camelcase
  default_error_message: 'The request cannot be processed. Try again later.'
};

let allMessages: Messages = {};
let locale: string | undefined;

export function translate(...keys: string[]): string {
  const messageKey = keys.join('.');
  const l10nMessages = getMessages();
  if (process.env.NODE_ENV === 'development' && !l10nMessages[messageKey]) {
    // eslint-disable-next-line no-console
    console.error(`No message for: ${messageKey}`);
  }
  return l10nMessages[messageKey] || messageKey;
}

export function translateWithParameters(
  messageKey: string,
  ...parameters: Array<string | number>
): string {
  const message = getMessages()[messageKey];
  if (message) {
    return parameters
      .map(parameter => String(parameter))
      .reduce((acc, parameter, index) => acc.replace(`{${index}}`, () => parameter), message);
  }
  if (process.env.NODE_ENV === 'development') {
    // eslint-disable-next-line no-console
    console.error(`No message for: ${messageKey}`);
  }
  return `${messageKey}.${parameters.join('.')}`;
}

export function hasMessage(...keys: string[]): boolean {
  const messageKey = keys.join('.');
  return getMessages()[messageKey] != null;
}

export function getMessages() {
  if (typeof allMessages === 'undefined') {
    logWarning('L10n messages are not initialized. Use default messages.');
    return DEFAULT_MESSAGES;
  }
  return allMessages;
}

export function resetMessages(newMessages: Messages) {
  allMessages = newMessages;
}

export function getLocalizedMetricName(
  metric: { key: string; name?: string },
  short = false
): string {
  const bundleKey = `metric.${metric.key}.${short ? 'short_name' : 'name'}`;
  if (hasMessage(bundleKey)) {
    return translate(bundleKey);
  } else if (short) {
    return getLocalizedMetricName(metric);
  }
  return metric.name || metric.key;
}

export function getLocalizedCategoryMetricName(metric: { key: string; name?: string }) {
  const bundleKey = `metric.${metric.key}.extra_short_name`;
  return hasMessage(bundleKey) ? translate(bundleKey) : getLocalizedMetricName(metric, true);
}

export function getLocalizedMetricDomain(domainName: string) {
  const bundleKey = `metric_domain.${domainName}`;
  return hasMessage(bundleKey) ? translate(bundleKey) : domainName;
}

export function getCurrentLocale() {
  return locale;
}

export function resetCurrentLocale(newLocale: string) {
  locale = newLocale;
}

export function getShortMonthName(index: number) {
  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec'
  ];
  return translate(months[index]);
}

export function getWeekDayName(index: number) {
  const weekdays = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  return weekdays[index] ? translate(weekdays[index]) : '';
}

export function getShortWeekDayName(index: number) {
  const weekdays = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  return weekdays[index] ? translate(weekdays[index]) : '';
}

const L10N_BUNDLE_LS_KEY = 'l10n.bundle';

export async function loadL10nBundle() {
  const bundle = await getLatestL10nBundle().catch(() => ({
    locale: DEFAULT_LOCALE,
    messages: {}
  }));

  resetCurrentLocale(bundle.locale);
  resetMessages(bundle.messages);

  // No need to load english (default) bundle, it's coming with react-intl
  if (bundle.locale !== DEFAULT_LOCALE) {
    const [intlBundle, intl] = await Promise.all([
      import(`react-intl/locale-data/${bundle.locale}`),
      import('react-intl')
    ]);

    intl.addLocaleData(intlBundle.default);
  }

  return bundle;
}

export async function getLatestL10nBundle() {
  const browserLocale = getPreferredLanguage();
  const cachedBundle = loadL10nBundleFromLocalStorage();

  const params: L10nBundleRequestParams = {};

  if (browserLocale) {
    params.locale = browserLocale;

    if (
      cachedBundle.locale &&
      browserLocale.startsWith(cachedBundle.locale) &&
      cachedBundle.timestamp &&
      cachedBundle.messages
    ) {
      params.ts = cachedBundle.timestamp;
    }
  }

  const { effectiveLocale, messages } = await fetchL10nBundle(params).catch(response => {
    if (response && response.status === 304) {
      return {
        effectiveLocale: cachedBundle.locale || browserLocale || DEFAULT_LOCALE,
        messages: cachedBundle.messages ?? {}
      };
    }
    throw new Error(`Unexpected status code: ${response.status}`);
  });

  const bundle = {
    timestamp: toNotSoISOString(new Date()),
    locale: effectiveLocale,
    messages
  };

  saveL10nBundleToLocalStorage(bundle);

  return bundle;
}

export function getCurrentL10nBundle() {
  return loadL10nBundleFromLocalStorage();
}

function getPreferredLanguage() {
  return window.navigator.languages ? window.navigator.languages[0] : window.navigator.language;
}

function loadL10nBundleFromLocalStorage() {
  let bundle: L10nBundle;

  try {
    bundle = JSON.parse(loadFromLocalStorage(L10N_BUNDLE_LS_KEY) ?? '{}');
  } catch {
    bundle = {};
  }

  return bundle;
}

function saveL10nBundleToLocalStorage(bundle: L10nBundle) {
  saveInLocalStorage(L10N_BUNDLE_LS_KEY, JSON.stringify(bundle));
}

function logWarning(message: string) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.warn(message);
  }
}
