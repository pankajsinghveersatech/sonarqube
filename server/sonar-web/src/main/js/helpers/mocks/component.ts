/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier, TreeComponent } from '../../types/component';
import { MetricKey } from '../../types/metrics';
import { mockMeasureEnhanced } from '../testMocks';

export function mockComponent(overrides: Partial<T.Component> = {}): T.Component {
  return {
    breadcrumbs: [],
    key: 'my-project',
    name: 'MyProject',
    qualifier: ComponentQualifier.Project,
    qualityGate: { isDefault: true, key: '30', name: 'Sonar way' },
    qualityProfiles: [
      {
        deleted: false,
        key: 'my-qp',
        language: 'ts',
        name: 'Sonar way'
      }
    ],
    tags: [],
    ...overrides
  };
}

export function mockTreeComponent(overrides: Partial<TreeComponent>): TreeComponent {
  return {
    key: 'my-key',
    qualifier: ComponentQualifier.Project,
    name: 'component',
    visibility: 'public',
    ...overrides
  };
}

export function mockComponentMeasure(
  file = false,
  overrides: Partial<T.ComponentMeasure> = {}
): T.ComponentMeasure {
  if (file) {
    return {
      key: 'foo:src/index.tsx',
      name: 'index.tsx',
      qualifier: ComponentQualifier.File,
      path: 'src/index.tsx',
      measures: [{ metric: MetricKey.bugs, value: '1', bestValue: false }],
      ...overrides
    };
  }
  return {
    key: 'foo',
    name: 'Foo',
    qualifier: ComponentQualifier.Project,
    measures: [{ metric: MetricKey.bugs, value: '12', bestValue: false }],
    ...overrides
  };
}

export function mockComponentMeasureEnhanced(
  overrides: Partial<T.ComponentMeasureEnhanced> = {}
): T.ComponentMeasureEnhanced {
  return {
    ...mockComponentMeasure(false, overrides as T.ComponentMeasure),
    leak: undefined,
    measures: [mockMeasureEnhanced()],
    value: undefined,
    ...overrides
  };
}
