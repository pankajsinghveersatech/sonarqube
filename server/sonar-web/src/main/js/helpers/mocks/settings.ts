/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Setting, SettingCategoryDefinition, SettingWithCategory } from '../../types/settings';

export function mockDefinition(
  overrides: Partial<SettingCategoryDefinition> = {}
): SettingCategoryDefinition {
  return {
    key: 'foo',
    category: 'foo category',
    fields: [],
    options: [],
    subCategory: 'foo subCat',
    ...overrides
  };
}

export function mockSetting(overrides: Partial<Setting> = {}): Setting {
  return {
    key: 'foo',
    value: '42',
    inherited: true,
    definition: {
      key: 'foo',
      name: 'Foo setting',
      description: 'When Foo then Bar',
      type: 'INTEGER',
      options: []
    },
    ...overrides
  };
}

export function mockSettingWithCategory(
  overrides: Partial<SettingWithCategory> = {}
): SettingWithCategory {
  return {
    key: 'foo',
    value: '42',
    inherited: true,
    definition: {
      key: 'foo',
      name: 'Foo setting',
      description: 'When Foo then Bar',
      type: 'INTEGER',
      options: [],
      category: 'general',
      fields: [],
      subCategory: 'email'
    },
    ...overrides
  };
}
