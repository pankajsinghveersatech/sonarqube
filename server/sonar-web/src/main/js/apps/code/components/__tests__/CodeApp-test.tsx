/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { mockPullRequest } from '../../../../helpers/mocks/branch-like';
import { mockComponent, mockComponentMeasure } from '../../../../helpers/mocks/component';
import { mockIssue, mockRouter } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import { ComponentQualifier } from '../../../../types/component';
import { loadMoreChildren, retrieveComponent } from '../../utils';
import { CodeApp } from '../CodeApp';

jest.mock('../../utils', () => ({
  loadMoreChildren: jest.fn().mockResolvedValue({}),
  retrieveComponent: jest.fn().mockResolvedValue({
    breadcrumbs: [],
    component: { qualifier: 'APP' },
    components: [],
    page: 0,
    total: 1
  }),
  retrieveComponentChildren: () => Promise.resolve()
}));

const METRICS = {
  coverage: { id: '2', key: 'coverage', type: 'PERCENT', name: 'Coverage', domain: 'Coverage' },
  new_bugs: { id: '4', key: 'new_bugs', type: 'INT', name: 'New Bugs', domain: 'Reliability' }
};

beforeEach(() => {
  (retrieveComponent as jest.Mock<any>).mockClear();
});

it.each([
  [ComponentQualifier.Application],
  [ComponentQualifier.Project],
  [ComponentQualifier.Portfolio],
  [ComponentQualifier.SubPortfolio]
])('should render correclty when no sub component for %s', async qualifier => {
  const component = { breadcrumbs: [], name: 'foo', key: 'foo', qualifier };
  (retrieveComponent as jest.Mock<any>).mockResolvedValueOnce({
    breadcrumbs: [],
    component,
    components: [],
    page: 0,
    total: 1
  });
  const wrapper = shallowRender({ component });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  wrapper.instance().handleSearchResults([]);
  expect(wrapper).toMatchSnapshot('no search');
  (retrieveComponent as jest.Mock<any>).mockResolvedValueOnce({
    breadcrumbs: [],
    component,
    components: [mockComponent({ qualifier: ComponentQualifier.File })],
    page: 0,
    total: 1
  });
  wrapper.instance().loadComponent(component.key);
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot('with sub component');
});

it('should refresh branch status if issues are updated', async () => {
  const fetchBranchStatus = jest.fn();
  const branchLike = mockPullRequest();
  const wrapper = shallowRender({ branchLike, fetchBranchStatus });
  const instance = wrapper.instance();
  await waitAndUpdate(wrapper);

  instance.handleIssueChange(mockIssue());
  expect(fetchBranchStatus).toBeCalledWith(branchLike, 'foo');
});

it('should load more behave correctly', async () => {
  const component1 = mockComponent();
  const component2 = mockComponent();
  (retrieveComponent as jest.Mock<any>).mockResolvedValueOnce({
    breadcrumbs: [],
    component: mockComponent(),
    components: [component1],
    page: 0,
    total: 1
  });
  let wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  (loadMoreChildren as jest.Mock<any>).mockResolvedValueOnce({
    components: [component2],
    page: 0,
    total: 1
  });

  wrapper.instance().handleLoadMore();
  expect(wrapper.state().components).toContainEqual(component1);
  expect(wrapper.state().components).toContainEqual(component2);

  (retrieveComponent as jest.Mock<any>).mockRejectedValueOnce({});
  wrapper = shallowRender();
  await waitAndUpdate(wrapper);
  wrapper.instance().handleLoadMore();
  expect(wrapper.state().components).toBeUndefined();
});

it('should handle go to parent correctly', async () => {
  const router = mockRouter();
  (retrieveComponent as jest.Mock<any>).mockResolvedValueOnce({
    breadcrumbs: [],
    component: mockComponent(),
    components: [],
    page: 0,
    total: 1
  });
  let wrapper = shallowRender();
  await waitAndUpdate(wrapper);
  wrapper.instance().handleGoToParent();
  expect(wrapper.state().highlighted).toBeUndefined();

  const breadcrumb = { key: 'key2', name: 'name2', qualifier: ComponentQualifier.Directory };
  (retrieveComponent as jest.Mock<any>).mockResolvedValueOnce({
    breadcrumbs: [
      { key: 'key1', name: 'name1', qualifier: ComponentQualifier.Directory },
      breadcrumb
    ],
    component: mockComponent(),
    components: [],
    page: 0,
    total: 1
  });
  wrapper = shallowRender({ router });
  await waitAndUpdate(wrapper);
  wrapper.instance().handleGoToParent();
  expect(wrapper.state().highlighted).toBe(breadcrumb);
  expect(router.push).toHaveBeenCalledWith({
    pathname: '/code',
    query: { id: 'foo', line: undefined, selected: 'key1' }
  });
});

it('should handle select correctly', () => {
  const router = mockRouter();
  const wrapper = shallowRender({ router });
  wrapper.setState({ highlighted: mockComponentMeasure() });

  wrapper.instance().handleSelect(mockComponentMeasure(true, { refKey: 'test' }));
  expect(router.push).toHaveBeenCalledWith({
    pathname: '/dashboard',
    query: { branch: undefined, id: 'test' }
  });
  expect(wrapper.state().highlighted).toBeUndefined();

  wrapper.instance().handleSelect(mockComponentMeasure());
  expect(router.push).toHaveBeenCalledWith({
    pathname: '/dashboard',
    query: { branch: undefined, id: 'test' }
  });
});

function shallowRender(props: Partial<CodeApp['props']> = {}) {
  return shallow<CodeApp>(
    <CodeApp
      component={{
        breadcrumbs: [],
        name: 'foo',
        key: 'foo',
        qualifier: 'FOO'
      }}
      fetchBranchStatus={jest.fn()}
      fetchMetrics={jest.fn()}
      location={{ query: { branch: 'b', id: 'foo', line: '7' } }}
      metrics={METRICS}
      router={mockRouter()}
      {...props}
    />
  );
}
