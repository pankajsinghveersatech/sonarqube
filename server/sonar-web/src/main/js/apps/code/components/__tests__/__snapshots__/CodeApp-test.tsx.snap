// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correclty when no sub component for APP 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <div
    className="code-components"
  >
    <div
      className="display-flex-center display-flex-column no-file"
    >
      <span
        className="h1 text-muted"
      >
        code_viewer.no_source_code_displayed_due_to_empty_analysis.APP
      </span>
    </div>
  </div>
</div>
`;

exports[`should render correclty when no sub component for APP: no search 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <withRouter(Search)
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "foo",
        "name": "foo",
        "qualifier": "APP",
      }
    }
    onSearchClear={[Function]}
    onSearchResults={[Function]}
  />
  <div
    className="code-components"
  >
    <div
      className="boxed-group spacer-top search-results"
    >
      <withKeyboardNavigation(Components)
        components={Array []}
        metrics={Object {}}
        onHighlight={[Function]}
        onSelect={[Function]}
        rootComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "APP",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`should render correclty when no sub component for APP: with sub component 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <withRouter(Search)
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "foo",
        "name": "foo",
        "qualifier": "APP",
      }
    }
    onSearchClear={[Function]}
    onSearchResults={[Function]}
  />
  <div
    className="code-components"
  >
    <div
      className="boxed-group spacer-top"
    >
      <withKeyboardNavigation(Components)
        baseComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "APP",
          }
        }
        components={
          Array [
            Object {
              "breadcrumbs": Array [],
              "key": "my-project",
              "name": "MyProject",
              "qualifier": "FIL",
              "qualityGate": Object {
                "isDefault": true,
                "key": "30",
                "name": "Sonar way",
              },
              "qualityProfiles": Array [
                Object {
                  "deleted": false,
                  "key": "my-qp",
                  "language": "ts",
                  "name": "Sonar way",
                },
              ],
              "tags": Array [],
            },
          ]
        }
        cycle={true}
        metrics={
          Object {
            "coverage": Object {
              "domain": "Coverage",
              "id": "2",
              "key": "coverage",
              "name": "Coverage",
              "type": "PERCENT",
            },
            "new_bugs": Object {
              "domain": "Reliability",
              "id": "4",
              "key": "new_bugs",
              "name": "New Bugs",
              "type": "INT",
            },
          }
        }
        onEndOfList={[Function]}
        onGoToParent={[Function]}
        onHighlight={[Function]}
        onSelect={[Function]}
        rootComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "APP",
          }
        }
      />
    </div>
    <ListFooter
      count={1}
      loadMore={[Function]}
      total={1}
    />
  </div>
</div>
`;

exports[`should render correclty when no sub component for SVW 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <div
    className="code-components"
  >
    <div
      className="display-flex-center display-flex-column no-file"
    >
      <span
        className="h1 text-muted"
      >
        code_viewer.no_source_code_displayed_due_to_empty_analysis.SVW
      </span>
    </div>
  </div>
</div>
`;

exports[`should render correclty when no sub component for SVW: no search 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <withRouter(Search)
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "foo",
        "name": "foo",
        "qualifier": "SVW",
      }
    }
    onSearchClear={[Function]}
    onSearchResults={[Function]}
  />
  <div
    className="code-components"
  >
    <div
      className="boxed-group spacer-top search-results"
    >
      <withKeyboardNavigation(Components)
        components={Array []}
        metrics={Object {}}
        onHighlight={[Function]}
        onSelect={[Function]}
        rootComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "SVW",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`should render correclty when no sub component for SVW: with sub component 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <withRouter(Search)
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "foo",
        "name": "foo",
        "qualifier": "SVW",
      }
    }
    onSearchClear={[Function]}
    onSearchResults={[Function]}
  />
  <div
    className="code-components"
  >
    <div
      className="boxed-group spacer-top"
    >
      <withKeyboardNavigation(Components)
        baseComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "SVW",
          }
        }
        components={
          Array [
            Object {
              "breadcrumbs": Array [],
              "key": "my-project",
              "name": "MyProject",
              "qualifier": "FIL",
              "qualityGate": Object {
                "isDefault": true,
                "key": "30",
                "name": "Sonar way",
              },
              "qualityProfiles": Array [
                Object {
                  "deleted": false,
                  "key": "my-qp",
                  "language": "ts",
                  "name": "Sonar way",
                },
              ],
              "tags": Array [],
            },
          ]
        }
        cycle={true}
        metrics={
          Object {
            "coverage": Object {
              "domain": "Coverage",
              "id": "2",
              "key": "coverage",
              "name": "Coverage",
              "type": "PERCENT",
            },
            "new_bugs": Object {
              "domain": "Reliability",
              "id": "4",
              "key": "new_bugs",
              "name": "New Bugs",
              "type": "INT",
            },
          }
        }
        onEndOfList={[Function]}
        onGoToParent={[Function]}
        onHighlight={[Function]}
        onSelect={[Function]}
        rootComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "SVW",
          }
        }
      />
    </div>
    <ListFooter
      count={1}
      loadMore={[Function]}
      total={1}
    />
  </div>
</div>
`;

exports[`should render correclty when no sub component for TRK 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="code.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <div
    className="code-components"
  >
    <div
      className="display-flex-center display-flex-column no-file"
    >
      <span
        className="h1 text-muted"
      >
        code_viewer.no_source_code_displayed_due_to_empty_analysis.TRK
      </span>
    </div>
  </div>
</div>
`;

exports[`should render correclty when no sub component for TRK: no search 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="code.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <withRouter(Search)
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "foo",
        "name": "foo",
        "qualifier": "TRK",
      }
    }
    onSearchClear={[Function]}
    onSearchResults={[Function]}
  />
  <div
    className="code-components"
  >
    <div
      className="boxed-group spacer-top search-results"
    >
      <withKeyboardNavigation(Components)
        components={Array []}
        metrics={Object {}}
        onHighlight={[Function]}
        onSelect={[Function]}
        rootComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "TRK",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`should render correclty when no sub component for TRK: with sub component 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="code.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <withRouter(Search)
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "foo",
        "name": "foo",
        "qualifier": "TRK",
      }
    }
    onSearchClear={[Function]}
    onSearchResults={[Function]}
  />
  <div
    className="code-components"
  >
    <div
      className="boxed-group spacer-top"
    >
      <withKeyboardNavigation(Components)
        baseComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "TRK",
          }
        }
        components={
          Array [
            Object {
              "breadcrumbs": Array [],
              "key": "my-project",
              "name": "MyProject",
              "qualifier": "FIL",
              "qualityGate": Object {
                "isDefault": true,
                "key": "30",
                "name": "Sonar way",
              },
              "qualityProfiles": Array [
                Object {
                  "deleted": false,
                  "key": "my-qp",
                  "language": "ts",
                  "name": "Sonar way",
                },
              ],
              "tags": Array [],
            },
          ]
        }
        cycle={true}
        metrics={
          Object {
            "coverage": Object {
              "domain": "Coverage",
              "id": "2",
              "key": "coverage",
              "name": "Coverage",
              "type": "PERCENT",
            },
            "new_bugs": Object {
              "domain": "Reliability",
              "id": "4",
              "key": "new_bugs",
              "name": "New Bugs",
              "type": "INT",
            },
          }
        }
        onEndOfList={[Function]}
        onGoToParent={[Function]}
        onHighlight={[Function]}
        onSelect={[Function]}
        rootComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "TRK",
          }
        }
      />
    </div>
    <ListFooter
      count={1}
      loadMore={[Function]}
      total={1}
    />
  </div>
</div>
`;

exports[`should render correclty when no sub component for VW 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <div
    className="code-components"
  >
    <div
      className="display-flex-center display-flex-column no-file"
    >
      <span
        className="h1 text-muted"
      >
        code_viewer.no_source_code_displayed_due_to_empty_analysis.VW
      </span>
    </div>
  </div>
</div>
`;

exports[`should render correclty when no sub component for VW: no search 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <withRouter(Search)
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "foo",
        "name": "foo",
        "qualifier": "VW",
      }
    }
    onSearchClear={[Function]}
    onSearchResults={[Function]}
  />
  <div
    className="code-components"
  >
    <div
      className="boxed-group spacer-top search-results"
    >
      <withKeyboardNavigation(Components)
        components={Array []}
        metrics={Object {}}
        onHighlight={[Function]}
        onSelect={[Function]}
        rootComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "VW",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`should render correclty when no sub component for VW: with sub component 1`] = `
<div
  className="page page-limited"
>
  <Suggestions
    suggestions="code"
  />
  <Helmet
    defer={false}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <A11ySkipTarget
    anchor="code_main"
  />
  <withRouter(Search)
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "foo",
        "name": "foo",
        "qualifier": "VW",
      }
    }
    onSearchClear={[Function]}
    onSearchResults={[Function]}
  />
  <div
    className="code-components"
  >
    <div
      className="boxed-group spacer-top"
    >
      <withKeyboardNavigation(Components)
        baseComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "VW",
          }
        }
        components={
          Array [
            Object {
              "breadcrumbs": Array [],
              "key": "my-project",
              "name": "MyProject",
              "qualifier": "FIL",
              "qualityGate": Object {
                "isDefault": true,
                "key": "30",
                "name": "Sonar way",
              },
              "qualityProfiles": Array [
                Object {
                  "deleted": false,
                  "key": "my-qp",
                  "language": "ts",
                  "name": "Sonar way",
                },
              ],
              "tags": Array [],
            },
          ]
        }
        cycle={true}
        metrics={
          Object {
            "coverage": Object {
              "domain": "Coverage",
              "id": "2",
              "key": "coverage",
              "name": "Coverage",
              "type": "PERCENT",
            },
            "new_bugs": Object {
              "domain": "Reliability",
              "id": "4",
              "key": "new_bugs",
              "name": "New Bugs",
              "type": "INT",
            },
          }
        }
        onEndOfList={[Function]}
        onGoToParent={[Function]}
        onHighlight={[Function]}
        onSelect={[Function]}
        rootComponent={
          Object {
            "breadcrumbs": Array [],
            "key": "foo",
            "name": "foo",
            "qualifier": "VW",
          }
        }
      />
    </div>
    <ListFooter
      count={1}
      loadMore={[Function]}
      total={1}
    />
  </div>
</div>
`;
